docs/proposals/20191016-clusterctl-redesign.md

```diff
@@ -0,0 +1,265 @@
+
+At high level, the upgrade sequence consist of two operations:
+
+1. Delete all the provider components of the current version [3]
+1. Create provider components for the new version
+
+The new provider version will be then responsible for conversion of the related cluster API objects when necessary.
+
+*[1] Upgrading the management cluster and upgrading workload clusters are considered out of scope of clusterctl upgrades.*
+
+*[2] in case of more than one instance of the same provider is installed on the management cluster, all the instance of the same provider will be upgraded in a single operation because providers have a mix of namespaced objects and global objects, and thus it is not possible to fully isolate a provider versions.*
+
+*[3] TBD exact details with regards to provider CRDs in order to make object conversion possible*
+
+**clusterctl delete [provider]**
+
+Deleting a provider sequence consist of the following actions:
+Identify all the provider components for a provider using the labels applied by clusterctl during install
+Delete all the provider components [1] with the exception of CRD definitions [2]
+
+*[1] in case of more than one instance of the same provider is installed on the management cluster, only the provider components in the provider namespace are going to be deleted thus preserving global components required for the proper functioning of other instances.*
+*TBD: if delete force deleting workload cluster or preserve them / how to make this behavior configurable*
+
+*[2] The clusterctl tools always try to preserve what is actually running and deployed, in that case the Cluster API objects for workload clusters. The --hard flag can be used to force deletion of the Cluster API objects and CRD definitions, the --hard flag*
+
+**clusterctl reset**
+
+Reset sequence goal is to restore the management cluster to its initial state [1], and this basically will be implemented as hard deletion of all the installed providers
+
+*[1] TBD: we should make delete force deleting workload cluster or preserve them (or give an option to the user for this)*
+
+### Risks and Mitigations
+
+- R: Change in clusterctl behavior can disrupt some current users
+- M: Work with the community to refine this spec and take their feedback into account.
+- R: Precondition of having a working Kubernetes cluster might increase the adoption bar, especially for new users.
+- M: Document requirements and quick ways to get a working Kubernetes cluster either locally (e.g. kind) or in the cloud (e.g. GKE, EKS, etc.).
+
+## Alternatives
+
+## Upgrade Strategy
+
+Upgrades will be considered a first class feature to clusterctl, and fully supported.  At the time of this writing we recommend that the Kubernetes operator backup their management clusters with a tool such as velero prior to upgrade.  If a failure condition occurs, the operator can then restore the state of their management cluster using that tool.
```

Comment:
This section should be rewritten to address `clusterctl` upgrades, for example I have clusterctl v0.1.0 installed and I want to upgrade to v0.2.0